src/main/java/io/nats/client/MsgDeliveryPool.java

```diff
@@ -0,0 +1,174 @@
+/*
+ *  Copyright (c) 2017 Apcera Inc. All rights reserved. This program and the accompanying
+ *  materials are made available under the terms of the MIT License (MIT) which accompanies this
+ *  distribution, and is available at http://opensource.org/licenses/MIT
+ */
+
+package io.nats.client;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.locks.*;
+
+
+class MsgDeliveryPool {
+    private List<MsgDeliveryWorker> workers = null;
+    private int                     idx;
+    private boolean                 shutdown;
+
+    // Expand the pool of workers.
+    // This call has no effect for a size lower to 1 or if given size
+    // is lower or equal to current pool size.
+    synchronized void setSize(int size) {
+        if (size <= 0) {
+            return;
+        }
+        if (this.workers == null) {
+            this.workers = new ArrayList<MsgDeliveryWorker>(size);
+        }
+        // We support only expansion of the pool at this point
+        if (size > this.workers.size()) {
+            final int added = size-this.workers.size();
+            for (int i=0; i<added; i++) {
+                MsgDeliveryWorker w = new MsgDeliveryWorker();
+                w.start();
+                this.workers.add(w);
+            }
+        }
+    }
+
+    synchronized int getSize() {
+        return this.workers.size();
+    }
+
+    synchronized void assignDeliveryWorker(SubscriptionImpl sub) {
+        int idx = this.idx;
+        if (++this.idx >= this.workers.size()) {
+            this.idx = 0;
+        }
+        final MsgDeliveryWorker worker = this.workers.get(idx);
+        sub.setDeliveryWorker(worker);
+    }
+
+    synchronized void shutdown() {
+        if (this.shutdown) {
+            return;
+        }
+        this.shutdown = true;
+        for (int i=0; i<this.workers.size(); i++) {
+            final MsgDeliveryWorker w = this.workers.get(i);
+            w.shutdown();
+        }
+        this.workers.clear();
+    }
+}
+
+class MsgDeliveryWorker extends Thread {
+    private final Lock          mu       = new ReentrantLock();
+    private final List<Message> msgs     = new LinkedList<Message>();
+    private final Condition     cond     = mu.newCondition();
+    private boolean             inWait   = false;
```

Comment:
`inWait` should be `volatile` if you are reading and writing it from different threads - two different threads can have different values in the CPU cache, so this needs to go to main memory.

Reply:
Same as above. The inWait is used when posting the message, which is done under the worker's lock. It is just to save the need for a signal.